keystore/filesystem/key_export.go

```diff
@@ -0,0 +1,298 @@
+type KeyFileClassifier interface {
+	ClassifyExportedKey(path string) *ExportedKey
+}
+
+// ExportedKey describes a key that can be exported from key store.
+//
+// `Purpose` describes the purpose of this key.
+// This is one of the `Purpose...` constants exported by this module.
+//
+// `ID` is either client ID, or zone ID, or nil depending on the purpose.
+//
+// `*Path` fields will be empty when not applicable.
+// For example, symmetric keys will not have public or private parts,
+// and only public or private key of a key pair may be present.
+type ExportedKey struct {
+	Purpose string
+	ID      []byte
+
+	PublicPath    string
+	PrivatePath   string
+	SymmetricPath string
+}
+
+// Exported key purpose constants:
+const (
+	PurposeAuthenticationSymKey       = "auth_key"
+	PurposePoisonRecordKeyPair        = "poison_key"
+	PurposeStorageClientKeyPair       = "storage"
+	PurposeStorageZoneKeyPair         = "zone"
+	PurposeTransportConnectorKeyPair  = "connector"
+	PurposeTransportTranslatorKeyPair = "translator"
+	PurposeTransportServerKeyPair     = "server"
+)
+
+// ExportPublicKey loads a public key for export.
+func (store *KeyStore) ExportPublicKey(key ExportedKey) (*keys.PublicKey, error) {
+	if key.PublicPath == "" {
+		return nil, nil
+	}
+	// This is getPublicKeyByFilename() but without cache thrashing.
+	return utils.LoadPublicKey(key.PublicPath)
+}
+
+// ExportPrivateKey loads a private key for export.
+func (store *KeyStore) ExportPrivateKey(key ExportedKey) (*keys.PrivateKey, error) {
+	if key.PrivatePath == "" {
+		return nil, nil
+	}
+	// This is getPrivateKeyByFilename() but without cache thrashing.
+	privateKey, err := utils.LoadPrivateKey(key.PrivatePath)
+	if err != nil {
+		return nil, err
+	}
+	decryptedKey, err := store.encryptor.Decrypt(privateKey.Value, key.ID)
+	if err != nil {
+		return nil, err
+	}
+	privateKey.Value = decryptedKey
+	return privateKey, nil
+}
+
+// ExportKeyPair loads a key pair for export.
+func (store *KeyStore) ExportKeyPair(key ExportedKey) (*keys.Keypair, error) {
+	publicKey, err := store.ExportPublicKey(key)
+	if err != nil {
+		return nil, err
+	}
+	privateKey, err := store.ExportPrivateKey(key)
+	if err != nil {
+		return nil, err
+	}
+	return &keys.Keypair{Public: publicKey, Private: privateKey}, nil
+}
+
+// ExportSymmetricKey loads a symmetric key for export.
+func (store *KeyStore) ExportSymmetricKey(key ExportedKey) ([]byte, error) {
+	if key.SymmetricPath == "" {
+		return nil, nil
+	}
+	// Yes, they are not encrypted.
+	return utils.ReadFile(key.SymmetricPath)
+}
+
+// DefaultKeyFileClassifier is a KeyFileClassifier for standard key types.
+type DefaultKeyFileClassifier struct{}
+
+// EnumerateExportedKeys prepares a list of keys that can be exported.
+// The keys are classified with default key file classifier.
+func (store *KeyStore) EnumerateExportedKeys() ([]ExportedKey, error) {
+	return store.EnumerateExportedKeysByClass(&DefaultKeyFileClassifier{})
```

Comment:
Mmm... It seems like we wouldn't need any state here, and there will be at most two implementations, for CE and EE respectively. However, I'm not sure what do mean with package variables. Could you please elaborate?

I've intended for user code (i.e., migration tool) to only call `EnumerateExportedKeys()`. This other method `EnumerateExportedKeysByClass()` will be used by Acra EE code to implement `EnumerateExportedKeys()` for its basic key store, passing an extended classifier. Migration tool in Acra EE will still use `EnumerateExportedKeys()` (this time for EE key store).

Reply:
to avoid each time allocation of empty struct (I dont know how golang actually compile it, but can predict that at least some pointer it may allocate on stack/heap), we can allocate it once on package level and use it every time when need (like singleton). For example how it used in flag package - https://golang.org/src/flag/flag.go#L1010
```
var defaultClassifier = &DefaultKeyFileClassifier{};

func (store *KeyStore) EnumerateExportedKeys() ([]ExportedKey, error) {
	return store.EnumerateExportedKeysByClass(defaultClassifier)
}
```

Reply:
Oh, I’ve misunderstood your original comment. I thought you meant to use global variables as a way to configure the “classifier” being used by default, with the idea that Acra EE will somehow set a new value for this value, substituting its own classifier.

But you seem to have meant the allocation optimization as above. I’ve updated the code to include it like you’ve written.